File: leetcode/cpp/RemoveDuplicatesFromSortedArray_II.cpp
//
// http://www.leetcode.com/onlinejudge#
// @Author: Xin Chen
// @Created on: 12/24/2012
// @Last modified: 12/24/2012
//

class Solution {
public:
    int removeDuplicates(int A[], int n) {
        if (n <= 1) return n;        
        
        int v = A[0], k = 0, ct = 1;
        for (int i = 1; i < n; i ++) {
            if (A[i] == v) {
                if (ct == 1) {
                    k ++;
                    if (k < i) A[k] = A[i];
                }
                ct ++;                
            } else {
                v = A[i];
                k ++;
                if (k < i) A[k] = A[i];
                ct = 1;
            } 
        }
                
        return k + 1;
    }    
};

// This works too, slightly better code.

class Solution2 {
public:
    int removeDuplicates(int A[], int n) {
        if (n <= 1) return n;        
        
        int v = A[0], k = 1, ct = 1;
        for (int i = 1; i < n; i ++) {
            if (A[i] == v) {
                ++ ct;                
                if (ct == 2) {
                    if (k < i) A[k] = A[i];
                    ++ k;
                }
            } else {
                v = A[i];
                if (k < i) A[k] = A[i];
                ++ k;
                ct = 1;
            } 
        }
                
        return k;
    }    
};


/*
Problem:

Follow up for "Remove Duplicates":
What if duplicates are allowed at most twice?

For example,
Given sorted array A = [1,1,1,2,2,3],

Your function should return length = 5, and A is now [1,1,2,2,3].

 */

File: leetcode_java/RemoveDuplicatesFromSortedArray_II.java
public class Solution {
    public int removeDuplicates(int[] nums) {
        int len = nums.length;
        if (len <= 1) return len;
        
        int v = nums[0], L = 1, ct = 1;
        for (int i = 1; i < len; ++ i) {
            if (v == nums[i]) {
                if (ct == 1) {
                    nums[L] = nums[i];
                    ++ L;
                    ++ ct; 
                }
            }
            else {
                v = nums[L] = nums[i];
                ++ L;
                ct = 1;
            }
        }
        
        return L;
    }
}

/**
Remove Duplicates from Sorted Array II
Difficulty: Medium

Follow up for "Remove Duplicates":
What if duplicates are allowed at most twice?

For example,
Given sorted array nums = [1,1,1,2,2,3],

Your function should return length = 5, with the first five elements 
of nums being 1, 1, 2, 2 and 3. It doesn't matter what you leave 
beyond the new length.
 */
